From 60365fa2e1f55e51b44788d5eac68c6ddce77266 Mon Sep 17 00:00:00 2001
From: Andrew Eikum <aeikum@codeweavers.com>
Date: Wed, 27 Dec 2017 13:31:59 -0600
Subject: [PATCH] HACK: steam: wine.inf: Add required Steam registry entries.

---
 loader/wine.inf.in | 25 ++++++++++++++++++++++---
 1 file changed, 22 insertions(+), 3 deletions(-)

diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index 5f9f61e2535..43cae2a2bc2 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -70,7 +70,8 @@ AddReg=\
     Printing,\
     Services, \
     SessionMgr,\
-    ThemeManager
+    ThemeManager, \
+    SteamClient
 
 [DefaultInstall]
 Needs=BaseInstall
@@ -97,7 +97,8 @@ AddReg=\
 AddReg=\
     CurrentVersionWow64.ntamd64,\
     VCRuntime.ntamd64,\
-    VersionInfo
+    LicenseInformation, \
+    SteamClient.ntamd64
 
 [DefaultInstall.ntarm64]
 Needs=BaseInstall
@@ -119,6 +119,7 @@ AddReg=\
     MCI,\
     Misc,\
     Packages.wow64,\
-    VersionInfo
+    VersionInfo, \
+    SteamClient.ntamd64
 
 [Wow64Install.ntx86]
 Needs=BaseWow64Install
@@ -126,7 +126,8 @@ AddReg=\
 AddReg=\
     CurrentVersionWow64.ntx86,\
     VCRuntime.ntamd64,\
-    VCRuntime.ntx86
+    VCRuntime.ntx86, \
+    SteamClient
 
 [Wow64Install.ntarm]
 Needs=BaseWow64Install
@@ -3867,6 +3870,22 @@ l_intl.nls
 winehid.inf,"@%12%\winehid.sys,-1"
 wineusb.inf,"@%12%\wineusb.sys,-1"
 winexinput.inf,"@%12%\winexinput.sys,-1"
+
+[SteamClient]
+HKCU,Software\Valve\Steam,"SteamPath",,"%16422%\Steam"
+HKCU,Software\Valve\Steam,"SteamExe",,"%16422%\Steam\Steam.exe"
+HKCU,Software\Valve\Steam\ActiveProcess,"PID",0x10001,0x0000fffe
+HKCU,Software\Valve\Steam\ActiveProcess,"SteamClientDll",,"%16422%\Steam\steamclient.dll"
+HKCU,Software\Valve\Steam\ActiveProcess,"SteamPath",,"%16422%\Steam"
+
+[SteamClient.ntamd64]
+HKCU,Software\Valve\Steam,"SteamPath",,"%16422%\Steam"
+HKCU,Software\Valve\Steam,"SteamExe",,"%16422%\Steam\Steam.exe"
+HKCU,Software\Valve\Steam\ActiveProcess,"PID",0x10001,0x0000fffe
+HKCU,Software\Valve\Steam\ActiveProcess,"SteamClientDll",,"%16426%\Steam\steamclient.dll"
+HKCU,Software\Valve\Steam\ActiveProcess,"SteamClientDll64",,"%16426%\Steam\steamclient64.dll"
+HKCU,Software\Valve\Steam\ActiveProcess,"SteamPath",,"%16426%\Steam"
+HKLM,Software\Wow6432Node\Valve\Steam,"InstallPath",,"%16422%\Steam"
 
 [NlsFiles]
 c_037.nls
--
